### PR TITLE
Problem: memiavl snapshot and pruning leave corrupted snapshot

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- make snapshot taking and pruning working on temporary directory, assume atomicity of directory rename operation.

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

**Review by mmsqe — APPROVED** [review]

File: memiavl/db.go [hunk 1]
```diff
@@ -303,7 +303,13 @@ func (db *DB) pruneSnapshots() {
 			}
 
 			db.logger.Info("prune snapshot", "name", name)
-			if err := os.RemoveAll(filepath.Join(db.dir, name)); err != nil {
+			tmpPath := filepath.Join(db.dir, name+"-tmp")
+			if err := os.Rename(filepath.Join(db.dir, name), tmpPath); err != nil {
+				db.logger.Error("failed to prune snapshot", "err", err)
```
Comment (JayT106) [inline memiavl/db.go:308]:
```suggestion
				db.logger.Error("failed to move the snapshot to tmp file", "err", err)
```

File: memiavl/db_test.go [hunk 2]
```diff
@@ -224,6 +224,8 @@ func TestLoadVersion(t *testing.T) {
 	}
 	require.NoError(t, db.WaitAsyncCommit())
 
+	require.NoError(t, db.WaitAsyncCommit())
```
Comment (JayT106) [inline memiavl/db_test.go:227]: why check NoError twice?

File: memiavl/db_test.go [hunk 3]
```diff
@@ -224,6 +224,8 @@ func TestLoadVersion(t *testing.T) {
...
+
```
Comment (yihuang) [inline memiavl/db_test.go:228]:
```suggestion
```

File: memiavl/db_test.go (hunk 2 above)
Comment (yihuang) [reply memiavl/db_test.go]: good catch, that's because of merge, fixed.

**Pull request merged**